src/util/render.js

```diff
@@ -0,0 +1,320 @@
+ *
+ * @param  {HTMLElement}    element The element to map
+ * @param  {Boolean} isSVG   If true, node is within an SVG
+ * @return {{ domMap: Array, plainlySetInnerHTML?: boolean, innerHTML?: string }} A DOM tree map
+ */
+const createDOMMap = function (element, isChild = false, isSVG = false, type = 'template') {
+	const children = [...element.childNodes];
+
+	// If the list contains a comment which says "$$plainly-set-inner-html" it means
+	// that the all children of the element will be set as unsafe `innerHTML´
+	// which means that there won't be any diffing or fancy things
+	// but we will just mark it as a pure component and leave it be.
+	for (const child of children) {
+		if (child?.nodeType === NODE_TYPE_COMMENT && child.textContent === '$$plainly-set-inner-html') {
+			return {
+				domMap: [],
+				plainlySetInnerHTML: true,
+				innerHTML: element.innerHTML,
+			};
+		}
+	}
+
+	const domMap = children.map((node, index) => {
+		const details = {
+			content: node.childNodes && node.childNodes.length > 0 ? null : node.textContent,
+			attributes: node.nodeType !== 1 ? [] : getAttributes(node.attributes),
+			type:
+				node.nodeType === NODE_TYPE_TEXT
+					? 'text'
+					: node.nodeType === NODE_TYPE_COMMENT
+					? 'comment'
+					: node.tagName.toLowerCase(),
+			node: node,
+		};
+
+		details.isSVG = isSVG || details.type === 'svg';
+		details.children =
+			isChild && isCustomElement({ node }) ? { domMap: [] } : createDOMMap(node, true, details.isSVG);
+		return details;
+	});
+
+	return {
+		domMap,
+	};
+};
+
+const getStyleMap = function (styles) {
+	return styles.split(';').reduce(function (arr, style) {
+		if (style.trim().indexOf(':') > 0) {
+			const styleArr = style.split(':');
+			arr.push({
+				name: styleArr[0] ? styleArr[0].trim() : '',
+				value: styleArr[1] ? styleArr[1].trim() : '',
+			});
+		}
+		return arr;
+	}, []);
+};
+
+const removeStyles = function (elem, styles) {
+	styles.forEach(function (style) {
+		elem.style[style] = '';
+	});
+};
+
+const changeStyles = function (elem, styles) {
+	styles.forEach(function (style) {
+		elem.style[style.name] = style.value;
+	});
+};
+
+const diffStyles = function (elem, styles) {
+	// Get style map
+	const styleMap = getStyleMap(styles);
+
+	// Get styles to remove
+	const remove = Array.prototype.filter.call(elem.style, function (style) {
+		const findStyle = styleMap.find(function (newStyle) {
+			return newStyle.name === style && newStyle.value === elem.style[style];
+		});
+		return findStyle === undefined;
+	});
+
+	// Add and remove styles
+	removeStyles(elem, remove);
+	changeStyles(elem, styleMap);
+};
+
+const removeAttributes = function (elem, attributes) {
+	attributes.forEach(function (attribute) {
+		// If the attribute is a class, use className
+		// Else if it's style, remove all styles
+		// Otherwise, use removeAttribute()
+		if (attribute.attributeName === 'class') {
+			try {
+				// The class name is read only (e.g. for svg)
```

Comment:
@lukas-schardt Why this check? Couldn't "class" also be set/removed via "element.removeAttribute" ?